### PR TITLE
buildPerlPackage: deduplicate -I paths in shebangs

### PR DESCRIPTION
**Pull request by volth**
It makes [shebangs lines 3 times shorter](https://github.com/NixOS/nixpkgs/pull/55786#issuecomment-463842415)

It is not big an improvement, but [sometimes it is critical](https://github.com/NixOS/nixpkgs/pull/66446#issuecomment-520181977), for example on Darwin with its 512-byte limit

### CODE REVIEW

File: pkgs/development/perl-modules/generic/builder.sh [hunk 1]
```diff
@@ -4,7 +4,11 @@ PERL5LIB="$PERL5LIB${PERL5LIB:+:}$out/lib/perl5/site_perl"
 
 perlFlags=
 for i in $(IFS=:; echo $PERL5LIB); do
-    perlFlags="$perlFlags -I$i"
+    # exclude paths pointing to perl derivation (like /nix/store/abcdfghijklmnpqrsvwxyz0123456789-perl-5.28.2/lib/perl5/site_perl)
+    # they are needless pointing to correct perl, and source of subtle bugs pointing to another perl
+    if [[ ! $i =~ /[0123456789abcdfghijklmnpqrsvwxyz]{32}-perl-[0-9.]+/lib/perl5/site_perl$ ]] && [[ $perlFlags != *" -I$i"* ]]; then
```
Comment (FRidh) [inline pkgs/development/perl-modules/generic/builder.sh]: Maybe `${perl}/lib/perl5/site_perl` instead?

Comment (volth) [reply pkgs/development/perl-modules/generic/builder.sh]: Maybe, I need to check which `perl` is there during cross-compilation

Comment (volth) [reply pkgs/development/perl-modules/generic/builder.sh]: @FRidh, ok.
also a bug was found (regex syntax requires `sed -E`, not `sed -e`), so I fixed it and added additional logging what exactly is patched

**Comment by volth** [comment]
Oh, from a 2 lines small fix (the first commit https://github.com/NixOS/nixpkgs/pull/66458/commits/323b7f53d015c904079957efe9c71dd6c8cd12f) 1 it evolved to huge refactoring

File: pkgs/development/perl-modules/generic/builder.sh [hunk 2]
```diff
@@ -4,38 +4,39 @@ PERL5LIB="$PERL5LIB${PERL5LIB:+:}$out/lib/perl5/site_perl"
 
 perlFlags=
 for i in $(IFS=:; echo $PERL5LIB); do
-    perlFlags="$perlFlags -I$i"
+    if [[ $i == "$hostperl/lib/perl5/site_perl" ]] || [[ $i == "$hostperl/lib/perl5/site_perl/"* ]]; then
+        # exclude paths pointing to perl derivation (like /nix/store/abcdfghijklmnpqrsvwxyz0123456789-perl-5.28.2-aarch64/lib/perl5/site_perl/5.28.2)
+        # they are needless when pointing to correct perl, and source of subtle bugs when pointing to wrong perl
+        true
+    elif [[ $i =~ /[0123456789abcdfghijklmnpqrsvwxyz]{32}-perl-[0-9.]+(-[^/]+)?/lib/perl5/site_perl($|/) ]]; then
+        echo -e "unexpected perl $i on PERL5LIB where\n hostperl=$hostperl\n devperl =$devperl\n fullperl=$fullperl"
```
Comment (FRidh) [inline pkgs/development/perl-modules/generic/builder.sh]: use `disallowedReferences` in the builder instead?

Comment (volth) [reply pkgs/development/perl-modules/generic/builder.sh]: Hm, no.
1. It is difficult to build the list for `disallowedReferences`.
I am worrying of a case of leaking `pkgs.perl` (or `pkgs.perl.withPackages ....`, or `pkgsCross.*.perl.withPackages ....`) into packages targeting non-default perl (`perldevel` or `perl530`) more than about closure size. There has been such a bug few months ago.
The regex covers all the variety of `(pkgsCross.*|pkgsi686|)perl*(.withPackages|)` packages except `$hostperl` which is checked in the `if` above.
2. This assertion will trigger even if the wrong derivation path does not land to `$out`, it is a signal that `buildPerlPackage`'s contract is misused.

File: pkgs/development/perl-modules/generic/default.nix [hunk 3]
```diff
@@ -46,6 +46,10 @@ toPerlModule(stdenv.mkDerivation (
     builder = ./builder.sh;
     buildInputs = buildInputs ++ [ perl ];
     nativeBuildInputs = nativeBuildInputs ++ [ (perl.dev or perl) ];
+    
+    # environment variables passed to ./builder.sh
+    hostperl = perl;
+    devperl = perl.dev or perl;
```
Comment (FRidh) [inline pkgs/development/perl-modules/generic/default.nix]: `lib.getDev` ?

Comment (volth) [reply pkgs/development/perl-modules/generic/default.nix]: Actually, I'd rename this `.dev`.
It is not `dev` in the sense as there is `cairo.dev`, it is a hack used only for cross-compilation of perl packages.
In that `.dev`, there is no pkg-config and include files as one might expect, but a `buildPlatform`'s `perl` binary which uses `hostPlatform`'s libraries.
So far they are pure Perl and there is no `.so`, everything is fine.

Insisting that it is `.dev` up using lib functions for that is misleading.

**Comment by volth** [comment]
Two packages found which use some perl sources as test data for parsers.
Those files were affected by patching shebangs.

**Comment by volth** [comment]
let's merge?
so I will build the next cpan update with it

**Comment by volth** [comment]
stalled

**Pull request closed without merging**